### PR TITLE
Remove the kubevirt-config configMap

### LINKED ISSUE #6779: Remove kubevirt-config traces, since the configmap is deprecated for more than one year
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug

/kind enhancement

**What happened**:

We deprecated kubevirt-config more than a year ago: https://groups.google.com/g/kubevirt-dev/c/_mJP0RilTd4/m/xdrfk1JABgAJ.

As @nunnatsa pointed out, we still have it in the code.

**What you expected to happen**:

Let's remove it for 0.48.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by nunnatsa**
The configMap was declared as deprecated a while ago in favor of  the configuratoin in the KubeVirt custom resource.

Signed-off-by: Nahshon Unna-Tsameret <nunnatsa@redhat.com>

**What this PR does**:
This PR remove the configMap support for configuration

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6779

**Special notes for your reviewer**:

**Release note**:

```release-note
kubevirt-config configMap is no longer supported for KubeVirt configuration
```

### CODE REVIEW

**Comment by xpivarc** [comment]
@nunnatsa Please ping me once you think it's ready to review.

**Comment by nunnatsa** [comment]
> @nunnatsa Please ping me once you think it's ready to review.

Ping :smiley:

**Comment by nunnatsa** [comment]
> @nunnatsa Please ping me once you think it's ready to review.

@xpivarc - Ping :smiley:

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (rmohr) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:64]: nit: I think you can directly write `request.MustParse("800m")` and assign it instead of introducing a new variable.

**Review by rmohr — APPROVED** [review]
/lgtm
/approve

Looks awesome.

**Review by xpivarc — APPROVED** [review]
Just a thought. Is it worth throwing an event or alert if we see the config map?

**Comment by rmohr** [comment]
> Just a thought. Is it worth throwing an event or alert if we see the config map?

/hold

Not a bad thought. An event may be nice. @nunnatsa what do you think?

**Comment by nunnatsa** [comment]
> > Just a thought. Is it worth throwing an event or alert if we see the config map?
> 
> /hold
> 
> Not a bad thought. An event may be nice. @nunnatsa what do you think?

@rmohr, @xpivarc - some questions: first, when would you expect to see it? anytime KV is running? on boot time? For the first option, is it worth to read the CM all the time? and to constantly create events? For the second option, nothing if the CM is created after the boot.

HCO currently "reconciles" the deletion of this CM: https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/pkg/controller/operands/kubevirtConfigMap.go#L30

We can do something similar, but instead of removing the CM, to create an event.

Another question is about this PR: do you want to do it as part of this PR? I would merge it and create a new one for the event. What do you think?

**Comment by xpivarc** [comment]
> > > Just a thought. Is it worth throwing an event or alert if we see the config map?
> > 
> > 
> > /hold
> > Not a bad thought. An event may be nice. @nunnatsa what do you think?
> 
> @rmohr, @xpivarc - some questions: first, when would you expect to see it? anytime KV is running? on boot time? For the first option, is it worth to read the CM all the time? and to constantly create events? For the second option, nothing if the CM is created after the boot.

What is boot time? I would generally expect us to fire it only once or in some time intervals. (It also depends if we pick event or metric+alert) The one-time event can be ensured by annotation on cm or any other way.
> 
> HCO currently "reconciles" the deletion of this CM: https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/pkg/controller/operands/kubevirtConfigMap.go#L30

:+1:  This is great but we should still have this event in the core. 
> 
> We can do something similar, but instead of removing the CM, to create an event.
> 
> Another question is about this PR: do you want to do it as part of this PR? I would merge it and create a new one for the event. What do you think?

I would prefer to have it in this PR so we ensure it is in one release if it's possible. @rmohr  wdyt?

**Comment by nunnatsa** [comment]
@xpivarc, @rmohr  - I opened the #6798 PR on top of this PR. Does it make any sense?

**Comment by rmohr** [comment]
> @xpivarc, @rmohr - I opened the #6798 PR on top of this PR. Does it make any sense?

What you have there makes a lot of sense. There is some event suppression in place, to not flood the apiserver too much, but I wonder how often the event will appear. We may want to suppress it ourselves a little bit more if needed, but if it is very insistent and annoying, that makes in my option sense.

File: pkg/virt-operator/kubevirt.go
Comment (rmohr) [inline pkg/virt-operator/kubevirt.go:1178]: This should defintively be a `EventTypeWarning`.

**Comment by nunnatsa** [comment]
@xpivarc, @rmohr - could you please review again?

File: pkg/virt-operator/kubevirt.go
Comment (nunnatsa) [reply pkg/virt-operator/kubevirt.go]: done

File: pkg/virt-api/api.go
Comment (xpivarc) [inline pkg/virt-api/api.go:885]: This needs to be moved somewhere else otherwise we have regression.

File: pkg/virt-controller/watch/application.go
Comment (xpivarc) [inline pkg/virt-controller/watch/application.go:284]: Same here.

File: pkg/virt-api/api.go
Comment (nunnatsa) [reply pkg/virt-api/api.go:885]: Thanks for the review @xpivarc. Not sure I get this point. Why would we have regression? The CM is no longer in use.

Comment (nunnatsa) [reply pkg/virt-api/api.go:885]: Checked again. This informer is only watches the obsolete configmap. I don't understand how it will cause regression.

File: pkg/virt-controller/watch/application.go
Comment (nunnatsa) [reply pkg/virt-controller/watch/application.go:284]: Same here - why will it cause regression?

File: pkg/virt-api/api.go
Comment (xpivarc) [reply pkg/virt-api/api.go:885]: @nunnatsa  Sorry for my not exact comment. I meant to comment on the line below. The important part is `apiHealthVersion.Clear()` in the handler. We need to move it to another handler. Meaning we need a handler on another informer.

It is an optimization so we would not make a frequent API call to figure out if `kube-api` is available. This is most important in `virt-handler` as the count of them can be very high.

Comment (nunnatsa) [reply pkg/virt-api/api.go:885]: @xpivarc - thanks. got it - fixed.

File: pkg/virt-controller/watch/application.go
Comment (nunnatsa) [reply pkg/virt-controller/watch/application.go:284]: @xpivarc - done

File: tests/vmi.json
Comment (enp0s3) [inline tests/vmi.json:1]: @nunnatsa Hi. Is this a leftover or a required file?

**Review by enp0s3 — COMMENTED** [review]
Hi, going over for the first time, please see my question below.

File: tests/vmi.json
Comment (nunnatsa) [reply tests/vmi.json:1]: oops...

File: pkg/virt-operator/kubevirt_test.go
Comment (enp0s3) [inline pkg/virt-operator/kubevirt_test.go:2575]: @nunnatsa I think we need ` defer close` for the channel here

File: pkg/virt-operator/kubevirt.go
Comment (enp0s3) [inline pkg/virt-operator/kubevirt.go:1165]: @nunnatsa WDYT about using `sync.Once.Do` here? this function should be called only once during virt-operator init right? otherwise ignore my comment here

File: pkg/testutils/mock_config.go
Comment (enp0s3) [inline pkg/testutils/mock_config.go:25]: @nunnatsa Can you elaborate on this initialization? I see it wasn't there before

File: pkg/testutils/mock_config.go
Comment (enp0s3) [inline pkg/testutils/mock_config.go:79]: @nunnatsa same here about the initialization `clone.Status.Phase = "Deployed"`

**Review by enp0s3 — COMMENTED** [review]
Doing a second round, focused on some aspects, please see questions below

File: pkg/virt-operator/kubevirt.go
Comment (rmohr) [inline pkg/virt-operator/kubevirt.go:1159]: This ticker logic seems to be slihtly over-engineered. how about using `wait.Until`:


```go
	wait.Until(func() {
		logger.V(5).Info("read the kubevirt-config configMap. if exist, emmit an event")
		cm, err := c.clientset.CoreV1().ConfigMaps(c.operatorNamespace).Get(ctx, obsoleteCmName, getOpts)
		if err != nil {
			if errors.IsNotFound(err) {
				logger.V(5).Info("The obsolete kubevirt-config configMap could not be found. good.")
			} else {
				logger.Errorf("can't get the kubevirt-config configMap %v", err)
			}
		}
		logger.Warning("The obsolete kubevirt-config configMap still exists. Please remove it.")
		c.recorder.Eventf(cm, k8sv1.EventTypeWarning, obsoleteCMReason, msg)
	}, time.Minute * 10, stopChan)
```

File: pkg/virt-operator/application.go
Comment (rmohr) [inline pkg/virt-operator/application.go:409]: You don't have to stop anything. If we loose the election and we just exit asap. Even this `leaderGauge.Set(0)` very likely has 0 effect and could be removed.

**Review by rmohr — CHANGES_REQUESTED** [review]

File: pkg/virt-operator/kubevirt.go
Comment (nunnatsa) [reply pkg/virt-operator/kubevirt.go:1159]: I don't have a strong opinion about it, but as far as I understand it, ticker is the "standard" golang way to do cron-like tasks.

Comment (rmohr) [reply pkg/virt-operator/kubevirt.go:1159]: Yes, just make your life a little bit easier and use `wait.Util` (btw. check out the wait file, there are really nice nuances of ticker based actions implemented.

File: pkg/virt-operator/application.go
Comment (nunnatsa) [reply pkg/virt-operator/application.go:409]: But with the current design, the goroutine keeps running. 

The problem is that the ticker's channel is read only and can't be closed.

Comment (rmohr) [reply pkg/virt-operator/application.go:409]: you don't need a ticker. Just make stop channel (you don't even ever need to stop this as I wrote).

File: pkg/virt-operator/kubevirt.go
Comment (nunnatsa) [reply pkg/virt-operator/kubevirt.go:1165]: Changing to another implementation as suggested by @rmohr - then it should not happen.

File: pkg/virt-operator/application.go
Comment (nunnatsa) [reply pkg/virt-operator/application.go:409]: Can't I relay on the existing stop channel?

Comment (rmohr) [reply pkg/virt-operator/application.go:409]: Sure :+1:

File: pkg/testutils/mock_config.go
Comment (nunnatsa) [reply pkg/testutils/mock_config.go:25]: Right. Some unit test failed. These test were based on a configMap configuration, but in this PR they are using KV configuration. The problem is in this code https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-config/config-map.go#L599-L601
The new configuration for the test are ignored, because the `phase` field is not set. Many test that uses KV configurations set the phase field within the test. Here we're ding it in one place for all the test. See https://github.com/kubevirt/kubevirt/blob/931a48c4c326d5f8e5799bc8e8ad02d3933157bc/pkg/testutils/mock_config.go#L67 in the same file, where the same solution is already in use.

File: pkg/testutils/mock_config.go
Comment (nunnatsa) [reply pkg/testutils/mock_config.go:79]: @enp0s3 - same answer.

File: pkg/virt-operator/kubevirt.go
Comment (nunnatsa) [reply pkg/virt-operator/kubevirt.go]: done

File: pkg/virt-api/api.go
Comment (rmohr) [inline pkg/virt-api/api.go:887]: It is slightly weird that we use this informer. I guess it is better than nothing ...

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter_test.go
Comment (rmohr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter_test.go:70]: Nit: I guess we can drop the `KVConfig` now that we only have one function. `testutils.NewFakeClusterConfig()` would be sufficient now.

File: pkg/virt-controller/watch/application.go
Comment (rmohr) [inline pkg/virt-controller/watch/application.go:282]: Let's use this informer everywhere. It is the highest-level informer.

**Review by rmohr — COMMENTED** [review]
Just one nit. Let's use the same informer in all components for health checks.

**Comment by rmohr** [comment]
/lgtm
/approve

Thanks everyone for the reviews.

**Comment by nunnatsa** [comment]
@rmohr please dismiss the change request.

File: pkg/virt-operator/kubevirt.go
Comment (xpivarc) [inline pkg/virt-operator/kubevirt.go:1170]: Should this stop?

Comment (rmohr) [reply pkg/virt-operator/kubevirt.go:1170]: You mean the loop? I don't think so. Often people have no kubevirt config first and just the default settings and then they want to do their first modification and either post a configmap or change the CR. That is what we want to catch.

**Comment by rmohr** [comment]
> @rmohr please dismiss the change request.

Github change requests have no influence on prow. If that is what you meant. :)

File: pkg/virt-operator/kubevirt.go
Comment (xpivarc) [reply pkg/virt-operator/kubevirt.go:1170]: Yes, sorry I apparently didn't wake up yet.

**Pull request merged**
